docker/Dockerfile

```diff
@@ -1,88 +1,71 @@
-                    git \
-                    yarn \
-                    gmp=6.2.1-r2 \
-&& apk --no-cache add \
-                  file \
-                  nodejs \
-                  linux-headers \
-                  runit \
-                  ttf-freefont \
-                  libreoffice \
-                  redis \
-                  postgresql-client
-
-# add non-root user and group with alpine first available uid, 1000
-RUN addgroup -g 1000 -S appgroup \
-&& adduser -u 1000 -S appuser -G appgroup
+ENV APP_HOME /app
+WORKDIR $APP_HOME
 
-# create app directory in conventional, existing dir /usr/src
-RUN mkdir -p /usr/src/app && mkdir -p /usr/src/app/tmp
-WORKDIR /usr/src/app
+RUN apk --update-cache upgrade \
+ && apk --no-cache add \
+                   build-base \
+                   git \
+                   gmp=6.2.1-r2 \
+                   postgresql-dev \
+                   yarn
 
-######################
-# DEPENDENCIES START #
-######################
-# Env vars needed for dependency install and asset precompilation?? maybe not
 ENV RAILS_ENV production
 
-COPY Gemfile* ./
+COPY Gemfile Gemfile.lock $APP_HOME/
 
-# only install production dependencies,
-# build nokogiri using libxml2-dev, libxslt-dev
-# note: installs bundler version used in Gemfile.lock
-#
 RUN gem install bundler -v $(cat Gemfile.lock | tail -1 | tr -d " ") \
 && bundle config without test development devunicorn \
-&& bundle config build.nokogiri --use-system-libraries \
-&& bundle install
+&& bundle install --jobs 4 --retry 3 \
+&& rm -rf /usr/local/bundle/cache/*.gem \
+&& find /usr/local/bundle/gems/ -name "*.c" -delete \
+&& find /usr/local/bundle/gems/ -name "*.o" -delete
 
-COPY package.json yarn.lock ./
+COPY package.json yarn.lock $APP_HOME/
 RUN yarn install --frozen-lockfile --production --silent
 
-####################
-# DEPENDENCIES END #
-####################
-COPY . .
+COPY . $APP_HOME/
 
-# precompile assets, silently (as verbose)
 RUN SECRET_KEY_BASE=a-real-secret-key-is-not-needed-here \
 RAILS_ENV=production \
 bundle exec rails assets:precompile 2> /dev/null
 
-# tidy up installation
-RUN apk --update-cache upgrade && apk del build-dependencies
+RUN rm -rf node_modules
+
+# Stage 2: copies dependencies from the build stage and creates final image for running the application
+FROM base
+LABEL Organisation="Ministry of Justice"
+LABEL Service="Claim for Crown Court Defence"
+LABEL Contact="crowncourtdefence@digital.justice.gov.uk"
+
+RUN set -ex
+
+ENV RAILS_ENV production
+ENV APP_HOME /usr/src/app
+WORKDIR $APP_HOME
+
+RUN apk --update-cache upgrade \
+ && apk --no-cache add \
+                   file \
+                   libreoffice \
+                   linux-headers \
+                   nodejs \
+                   postgresql-client \
+                   redis \
+                   runit \
+                   ttf-freefont
+
+RUN addgroup -g 1000 -S appgroup \
+&& adduser -u 1000 -S appuser -G appgroup
```

Comment:
Maybe also reduce this users permissions to the files it needs to write to like log, tmp, db etc:
```
# non-root/appuser should own only what they need to
RUN chown -R appuser:appgroup log tmp db
```

This is a line we have in CDUI's dockerfile at the moment.

i'm guessing the line `--chown=appuser:appgroup ./app` means that the appuser has ownership of all the files under app

Reply:
We originally had the same line `RUN chown -R appuser:appgroup log tmp db` in CCCD. I removed it after reading that explicitly running `chown` has a big impact on resulting image size and replaced it with the command you highlighted, which sets permissions when copying the files across. You're right that the drawback of this approach is that it's more permissive but seems to be the recommended approach (eg https://blog.mornati.net/docker-images-and-files-chown).